### ISSUE TITLE
Processes not terminated

### ISSUE THREAD

**Issue by zbzhu99** [huawei-noah/SMARTS#895]
## **BUG REPORT**
**High Level Description**
I use SMARTS for training imitation learning agents with NGSIM dataset. When training is finished, SMARTS will leave some unterminated processes.

**SMARTS version**
0.4.15

**Steps to reproduce the bug**
I use the `multiprocessing` package to open multiple SMARTS instances for parallel sampling.

And I created each simulator using:

```python
self.smarts = SMARTS(
    agent_interfaces={},
    traffic_sim=None,
    visdom=None,
    envision=None,
    zoo_addrs=None,
)
```

**Resulting and expected behaviour**
There are multiple processes like `smarts/zoo/worker.py --port xxxxx` not terminated after the training stops.

**Error logs and screenshots**
<img width="357" alt="截屏2021-06-02 11 16 23" src="https://user-images.githubusercontent.com/32090492/120420050-2482e180-c396-11eb-95a2-b6a49313f8db.png">
 
**System information**
 
- Operating system & Environment [Please note if using Docker]: Ubuntu 16.04
- Python version [Please note if using Jupyter] 3.7.3
- SUMO version 1.8.0

**Comment by Adaickalavan** [comment]
Could you provide a minimum working `.py` file example which runs multiple SMARTS instances and leaves behind unclosed processes?

**Comment by zbzhu99** [comment]
Here is an example script for reproducing the issue:

```python
from multiprocessing.context import Process
from multiprocessing import Pipe, connection
import gym

from smarts.core.agent_interface import AgentInterface, AgentType
from smarts.core.agent import AgentSpec, Agent


class SimpleAgent(Agent):
    def act(self, obs):
        return "keep_lane"


agent_spec = AgentSpec(
    interface=AgentInterface.from_type(AgentType.Laner, max_episode_steps=None),
    agent_builder=SimpleAgent,
)

agent_specs = {
    "Agent-007": agent_spec,
    "Agent-008": agent_spec,
}


def create_smarts():
    def _init():
        env = gym.make(
            "smarts.env:hiway-v0",
            scenarios=["scenarios/loop"],
            agent_specs=agent_specs,
            headless=True,
        )
        return env
    return _init


def _worker(
    parent: connection.Connection,
    p: connection.Connection,
    env_fn,
) -> None:

    parent.close()
    env = env_fn()
    try:
        while True:
            try:
                cmd, data = p.recv()
            except EOFError:  # the pipe has been closed
                p.close()
                break
            if cmd == "reset":
                obs = env.reset()
                p.send(obs)
            else:
                p.close()
                raise NotImplementedError
    except KeyboardInterrupt:
        p.close()


parent_remote, child_remote = Pipe()
args = (
    parent_remote,
    child_remote,
    create_smarts(),
)
process = Process(target=_worker, args=args, daemon=True)
process.start()
child_remote.close()
parent_remote.send(["reset", None])
obs = parent_remote.recv()
print(obs)
```

From my test, the unclosed processes might be related to setting them as daemon processes. If the processes are set to `daemon=False`, then there will no longer be unclosed processes.

However, setting `daemon=True` is adopted by openai baselines multi-env implementation, and that is also used in tianshou. Since I use the multiprocess env of tianshou, I encountered the issue in my experiment.

**Comment by Adaickalavan** [comment]
Hi @zbzhu99 ,

Thank you for your informative minimum working example and analysis of the problem.

Please try the fix in the branch `apocalypse` and let us know whether the issue is resolved.

**Comment by zbzhu99** [comment]
From my test, the issue is resolved with `apocalypse` branch. Thank you.

**Comment by Adaickalavan** [comment]
Closed by pull request #924 .

**Comment by Adaickalavan** [comment]
Hi @Zbzhu99,

We have now created `ParallelEnv` class to simulate multiple SMARTS environments in parallel, with synchronous or asynchronous episodes, to collect observations at a faster rate.

See pull request #1087 for details.

You may consider using `ParallelEnv` class to run mulitple SMARTS for parallel rollouts.
